discovery/syncer.go

```diff
@@ -933,102 +951,96 @@ func (g *GossipSyncer) replyChanRangeQuery(query *lnwire.QueryChannelRange) erro
-		if isFinalChunk {
-			channelChunk = channelRange[numChansSent:]
-
-			log.Infof("GossipSyncer(%x): sending final chan "+
-				"range chunk, size=%v", g.cfg.peerPub[:],
-				len(channelChunk))
-		} else {
-			// Otherwise, we'll only send off a fragment exactly
-			// sized to the proper chunk size.
-			channelChunk = channelRange[numChansSent : numChansSent+g.cfg.chunkSize]
-
-			log.Infof("GossipSyncer(%x): sending range chunk of "+
-				"size=%v", g.cfg.peerPub[:], len(channelChunk))
-		}
-
-		// If we have any channels at all to return, then we need to
-		// update our pointers to the first and last blocks for each
-		// response.
-		if len(channelChunk) > 0 {
-			// If this is the first response we'll send, we'll point
-			// the first block to the first block in the query.
-			// Otherwise, we'll continue from the block we left off
-			// at.
-			if numChansSent == 0 {
-				firstBlockHeight = startBlock
-			} else {
-				firstBlockHeight = lastBlockHeight
-			}
+	// We'll send our response in a streaming manner, chunk-by-chunk. We do
+	// this as there's a transport message size limit which we'll need to
+	// adhere to. We also need to make sure all of our replies cover the
+	// expected range of the query.
+	sendReplyForChunk := func(channelChunk []lnwire.ShortChannelID,
+		firstHeight, lastHeight uint32, finalChunk bool) error {
 
-			// If this is the last response we'll send, we'll point
-			// the last block to the last block of the query.
-			// Otherwise, we'll set it to the height of the last
-			// channel in the chunk.
-			if isFinalChunk {
-				lastBlockHeight = endBlock
-			} else {
-				lastBlockHeight = channelChunk[len(channelChunk)-1].BlockHeight
-			}
-		}
-
-		// The number of blocks contained in this response (the total
-		// span) is the difference between the last channel ID and the
-		// first in the range. We add one as even if all channels
+		// The number of blocks contained in the current chunk (the
+		// total span) is the difference between the last channel ID and
+		// the first in the range. We add one as even if all channels
 		// returned are in the same block, we need to count that.
-		numBlocksInResp := lastBlockHeight - firstBlockHeight + 1
+		numBlocks := lastHeight - firstHeight + 1
+		complete := uint8(0)
+		if finalChunk {
+			complete = 1
+		}
 
-		// With our chunk assembled, we'll now send to the remote peer
-		// the current chunk.
-		replyChunk := lnwire.ReplyChannelRange{
+		return g.cfg.sendToPeerSync(&lnwire.ReplyChannelRange{
 			QueryChannelRange: lnwire.QueryChannelRange{
 				ChainHash:        query.ChainHash,
-				NumBlocks:        numBlocksInResp,
-				FirstBlockHeight: firstBlockHeight,
+				NumBlocks:        numBlocks,
+				FirstBlockHeight: firstHeight,
 			},
-			Complete:     0,
+			Complete:     complete,
 			EncodingType: g.cfg.encodingType,
 			ShortChanIDs: channelChunk,
+		})
+	}
+
+	// If we don't have any channels to send, send an empty reply.
+	if len(channelRanges) == 0 {
```

Comment:
this case looks unnecessary, the reply at the end should send an equivalent response i think? log message looks like it'd also be accurate

Reply:
Yeah I had a slightly different approach before I arrived at the current one that required this check and forgot to remove it.